Raven.Client.Lightweight/Document/DocumentStore.cs

```diff
@@ -825,6 +909,174 @@ public Task GetObserveChangesAndEvictItemsFromCacheTask(string database = null)
+        {
+            try
+            {
+                var commands = nagleData.SelectMany(x => x.SaveChangesData.Commands).ToList();
+                var batchOptions = GetMergedBatchOptions(nagleData.Select(x => x.SaveChangesData.Options));
+                var results = DatabaseCommands.Batch(commands, batchOptions);
+                var index = 0;
+
+                foreach (var data in nagleData)
+                {
+                    data.TaskCompletionSource.TrySetResult(results.Skip(index).Take(data.SaveChangesData.Commands.Count).ToArray());
+                    index += data.SaveChangesData.Commands.Count;
+                }
+            }
+            catch (Exception)
+            {
+                RunCommandsOneByOne(nagleData);
+            }
+        }
+
+        private void RunCommandsOneByOne(List<NagleData> nagleData)
+        {
+            foreach (var data in nagleData)
+            {
+                try
+                {
+                    var batchResults = DatabaseCommands.Batch(data.SaveChangesData.Commands, data.SaveChangesData.Options);
+                    if (batchResults == null)
+                        throw new InvalidOperationException("Cannot call Save Changes after the document store was disposed.");
+
+                    data.TaskCompletionSource.TrySetResult(batchResults);
+                }
+                catch (Exception exception)
+                {
+                    data.TaskCompletionSource.TrySetException(exception);
+                }
+            }
+        }
+
+        private static BatchOptions GetMergedBatchOptions(IEnumerable<BatchOptions> allOptions)
+        {
+            var batchOptions = new BatchOptions();
+            foreach (var option in allOptions)
+            {
+                if (option == null)
+                    continue;
+
+                batchOptions.WaitForReplicas |= option.WaitForReplicas;
+                batchOptions.NumberOfReplicasToWaitFor = Math.Max(batchOptions.NumberOfReplicasToWaitFor, option.NumberOfReplicasToWaitFor);
+                batchOptions.WaitForReplicasTimout =
+                    batchOptions.WaitForReplicasTimout > option.WaitForReplicasTimout
+                        ? batchOptions.WaitForReplicasTimout
+                        : option.WaitForReplicasTimout;
+                batchOptions.Majority |= option.Majority;
+                batchOptions.ThrowOnTimeoutInWaitForReplicas |= option.ThrowOnTimeoutInWaitForReplicas;
+
+                batchOptions.WaitForIndexes |= option.WaitForIndexes;
+                batchOptions.WaitForIndexesTimeout =
+                    batchOptions.WaitForIndexesTimeout > option.WaitForIndexesTimeout
+                        ? batchOptions.WaitForIndexesTimeout
+                        : option.WaitForIndexesTimeout;
+
+                batchOptions.ThrowOnTimeoutInWaitForIndexes |= option.ThrowOnTimeoutInWaitForIndexes;
+                if (option.WaitForSpecificIndexes != null)
```

Comment:
Here this is wrong. 
If we have a request that wait for specific indexes, and one that waits for all, we need to wait for all

Reply:
done